File: Dynamic-program/718-findLength/718.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution
{
public:
    int findLength(vector<int> &nums1, vector<int> &nums2)
    {
        int len = 0;
        vector<vector<int>> dp(nums1.size() + 1, vector<int>(nums2.size() + 1, 0));
        for (int i = 1; i <= nums1.size(); ++i) {
            for (int j = 1; j <= nums2.size(); ++j) {
                if (nums1[i - 1] == nums2[j - 1]) {
                    dp[i][j] = dp[i - 1][j - 1] + 1;
                    len = max(len, dp[i][j]);
                }
            }
        }
        return len;
    }
};

File: Dynamic-program/718-findLength/go/718.go
package main

import "fmt"

func findLength(nums1 []int, nums2 []int) int {
	length := 0
	d := make([][]int, len(nums1)+1)
	for i := range d {
		d[i] = make([]int, len(nums2)+1)
	}

	for i := 1; i <= len(nums1); i++ {
		for j := 1; j <= len(nums2); j++ {
			if nums1[i-1] == nums2[j-1] {
				d[i][j] = d[i-1][j-1] + 1
				length = max(length, d[i][j])
			}
		}
	}

	return length
}

func main() {
	nums1 := []int{1, 2, 3, 2, 1}
	nums2 := []int{3, 2, 1, 4, 7}
	fmt.Println(findLength(nums1, nums2))
}